### ISSUE TITLE
Tor stuck on "Connecting to directory server" using IPv6

### ISSUE THREAD

**Issue by ali-ht10** [mtigas/OnionBrowser#73]
Hi,
Tor doesn't connect at all when using **IPv6 network**. It's stuck on **Connecting to directory server**.

This was the console log, 
**BOOTSTRAP PROGRESS=5 TAG=conn_dir SUMMARY="Connecting to directory server" WARNING="Network is unreachable" REASON=NOROUTE**

**Comment by mtigas** [comment]
@ali-ht10 
That's odd; on my cellular network (T-Mobile) it seems to connect fine, although that's possibly because I get both a "real" IPv4 and IPv6 address.

Can you try using the following custom bridge config lines:

```
[2605:6400:2:fed5:22:0:4d5d:7061]:443 5006A6C6891B8A4D16BB0B753F3144662678F2F6
[2001:41d0:8:b4cc::1]:44510 49C902DA4B1C512F67654975AFC253CEFD848DB3
```

Let me know if that works.

Unfortunately, IPv6 support in Tor is still somewhat limited; [software supports it pretty well now](https://trac.torproject.org/projects/tor/wiki/org/roadmaps/Tor/IPv6), but there are not many relays or bridges which support it. Making matters worse, current versions of Onion Browser do not support Pluggable Transport bridges, further limiting the number of available IPv6 bridges you might be able to use.

You can _try_ to get more bridge config lines here https://bridges.torproject.org/options but you will need to Use the **Advanced Options** section:

* **Do you need a Pluggable Transport?**: none
* **Do you need IPv6 addresses?**: Yes

Even then, I only rarely have success in getting IPv6 bridges because there are so few available.

**Comment by chrisballinger** [comment]
Perhaps this is related to the NAT64/DNS64 thing in iOS 9.3.3? If your
cellular carrier only supports IPv6 the system will synthesize both an IPv6
address as well as also returning a non-functioning IPv4 (using
getaddrinfo). The trick is to prefer the v6 address when available, but
then it causes problems when v6 is unreachable, so to truly fix the problem
you need to attempt both v6 and v4 and then pick the winner.

**Comment by ali-ht10** [comment]
My ISP doesn't provide IPv6. So I had to setup local IPv6 using "Internet Sharing" option in my Mac. Do you think that could be the issue that it's not connecting?

**Comment by ali-ht10** [comment]
And I did try to edit torrc file by using a combination of **ClientUseIPv6, ClientUseIPv4, ClientPreferIPv6DirPort, ClientPreferIPv6ORPort** but unfortunately no luck.

**Comment by ali-ht10** [comment]
@mtigas 
I changed the torrc file to this, 
**UseBridges 1
bridge [2605:6400:2:fed5:22:0:4d5d:7061]:443 5006A6C6891B8A4D16BB0B753F3144662678F2F6
bridge [2001:41d0:8:b4cc::1]:44510 49C902DA4B1C512F67654975AFC253CEFD848DB3**,
Tor still didn't connect.

Do you think I also have to include the above mentioned, **ClientUseIPv6, ClientUseIPv4, ClientPreferIPv6DirPort, ClientPreferIPv6ORPort** along with these custom bridges?

And I tried to get more bridges, but this was the message, **Uh oh, spaghettios! 
There currently aren't any bridges available... Perhaps you should try going back and choosing a different bridge type!**

**Comment by mtigas** [comment]
@ali-ht10 
Can you check http://test-ipv6.com/ and report the result that you get there? Curious to see what you get with your ISP and your local IPv6 sharing setting.

**Comment by ali-ht10** [comment]
No IPv6 address detected

Good news! Your current configuration will continue to work as web sites enable IPv6.

You appear to be able to browse the IPv4 Internet only. You will not be able to reach IPv6-only sites.

Your DNS server (possibly run by your ISP) appears to have IPv6 Internet access.

**Comment by JRock007** [comment]
Apple experienced this issue too :

> We discovered one or more bugs in your app when reviewed on iPad and iPhone running iOS 9.3.3 on Wi-Fi connected to an IPv6 network. Specifically, your app did not load beyond the "5% - Connecting to directory server" screen at launch.

I didn't get any more details unfortunately.

**Comment by ali-ht10** [comment]
I downloaded the Tor Browser from Tor's website and tried to use it using local IPv6. But it failed to connect.

**Comment by mtigas** [comment]
@ali-ht10 
It seems that your "local IPv6" setup does not allow you to actually _use_ IPv6 websites. ("No IPv6 address detected")

I get this with my ISP, which gives me both an IPv4 and IPv6 address:

```
Your IPv4 address on the public Internet appears to be xxx

Your IPv6 address on the public Internet appears to be xxx

Your Internet Service Provider (ISP) appears to be T-MOBILE-AS21928 - T-Mobile USA, Inc., US

Since you have IPv6, we are including a tab that shows how well you can reach other IPv6 sites. [more info]

Good news! Your current configuration will continue to work as web sites enable IPv6.

Your DNS server (possibly run by your ISP) appears to have IPv6 Internet access.

Your readiness score
10/10
```

On this ISP, I have success using Tor normally, and with the two IPv6 bridges I mentioned [above](https://github.com/mtigas/iOS-OnionBrowser/issues/73#issuecomment-235339966).

---

If you see http://test-ipv6.com/mirrors.html and you see that all pages have "X" in the "ipv6?" column, then that means that you are actually connecting to the IPv4 version of those domains — you only have "real" IPv6 connectivity if you are seeing a green ✓ in that column.

I'm not sure what you are doing for "local IPv6", but if you don't actually tunnel to the public IPv6 space, then you _definitely_ don’t have actual access to other IPv6 servers.

**Comment by mtigas** [comment]
@JRock007 
Per https://trac.torproject.org/projects/tor/wiki/org/roadmaps/Tor/IPv6 , the tor client previously only bootstrapped the initial connection on IPv4. The tor 0.2.8.X versions support this, but still don’t prefer IPv6 it seems (unless `ClientPreferIPv6ORPort 1` and `ClientPreferIPv6DirPort 1` are set).

In 0.2.8.X things _should_ just work (without the extra settings), but let me know if you’re still getting feedback like this from Apple review.

For this part of the thread (re: that Apple review comment) I wonder if we’re having issues with IPv6-only networks that give you non-functioning IPv4, like @chrisballinger mentioned.

**Comment by JRock007** [comment]
@mtigas I was already using a 0.2.8.X version of Tor, but I updated to 0.2.8.6 just in case.
Apple still reports the same issue, event with the ClientPreferIPv6ORPort 1 and ClientPreferIPv6DirPort 1 settings.

**Comment by dolar2006** [comment]
there is a solution to this problem?

**Comment by Drag11** [comment]
Any updates on that?

**Comment by mtigas** [comment]
Looked into this some more; had a conversation at a Tor dev meeting in September about the right way to get this working. Essentially, if you're in IPv6-only mode, you must set `ClientUseIPv4 0` and `ClientUseIPv6 1` to totally disable attempting connections via IPv4. We have to catch this situatino in the app side; this does not happen in Tor.

Anyway, I’ll have an Onion Browser beta over the weekend and (hopefully) a release next week that should allow proper support if the network only exposes IPv6.

Making matters worse, there is a current Tor bug where the app will not finish bootstrapping under IPv6-only unless you also disable microdescriptors. I've reported it and it’s currently being tracked upstream at: https://trac.torproject.org/projects/tor/ticket/20996 The workaround there makes the initial bootstrap very, very slow — but at least the app successfully works eventually. (The update only uses the slow workaround if the user has selected "IPv6 Only" or if the app is left in the default autodetect mode and the app has detected an IPv6-only network environment.)

I’ve noticed that T-Mobile US is now doing that (IPv6 external address, only loopback or utun IPv4 addresses), at least on my device. So it’s likely that this is going to start affecting a **lot** more users soon.

**Comment by mtigas** [comment]
Fixed as of 1.7.3.

* There’s [some detection](https://github.com/mtigas/OnionBrowser/blob/f4d185c36051337ccbcd14081e23898661feb59a/OnionBrowser/OnionBrowser/AppDelegate.m#L481-L496) of IPv6-only (so that we set the proper `ClientUseIPv4`/`ClientUseIPv6` flags)
* And the app is now using a 0.3.0.X alpha version of Tor which includes the fix from [the previously mentioned Tor ticket](https://trac.torproject.org/projects/tor/ticket/20996).
